### PR TITLE
Load from S3

### PR DESCRIPTION
**Pull request by jburel**
* Add notebook loading metadata from IDR and binary from S3
* Add corresponding Python script

The version of ``s3fs`` installed when using Python 2 is ``0.2.x``. An ``Access denied`` error is then thrown when attempting to load data from ``s3.embassy``. It is possible when using ``0.3.0+`` (current version is ``0.4.2``), unfortunately those versions are not available for Python 2.

### CODE REVIEW

**Comment by jburel** [comment]
The version of ``s3fs`` installed when using Python 2 is ``0.2.x``. An ``Access denied`` error is then thrown when attempting to load data from ``s3.embassy``. It is possible when using ``0.3.0+`` (current version is ``0.4.2``), unfortunately those versions are not available for Python 2.

File: notebooks/idr0002_zarr.ipynb [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+    "# Cell/particle Counting and scoring stained objects"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This notebook demonstrates how to process plates associated to the paper ['Integration of biological data by kernels on graph nodes allows prediction of new genes involved in mitotic chromosome condensation.'](http://dx.doi.org/10.1091/mbc.E13-04-0221) using [CellProfiler](http://cellprofiler.org/).\n",
+    "We use the example pipeline [Cell/particle counting, and scoring the percentage of stained objects](http://cellprofiler.org/examples/#PercentPositive). This pipeline is for two channels images.\n",
+    "Metadata are loaded from IDR and binary data from S3"
```
Comment (pwalczysko) [inline notebooks/idr0002_zarr.ipynb:16]: ``two channels images`` -> ``two-channel images``

**Comment by pwalczysko** [comment]
repo2docker

Executed cell of the notebook ``idr0002_zarr`` is

 *Import packages*

```
/srv/conda/envs/kernel/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
```
slightly annoying output

**Comment by jburel** [comment]
@pwalczysko it might be fixed in CP 4 :-)

**Comment by pwalczysko** [comment]
repo2docker

cell 
*Load from S3*

this was really fast, but there was no output saying how fast that was. Not sure what is the comparison here - I understand that we have this workflow in our training portfolio, but somehow a speedup from 3 sec to, say 0.001 sec is very hard to appreciate with only "feeling and memory". Both will be remembered as quite fast. I would also go through 5 wells, not just 2. Is the CP pipeline on the cells also somehow helped by the zarr format timewise ? I remember that being quite slow....

**Comment by pwalczysko** [comment]
repo2docker

```
conn.close()

WARNING:omero.client:Cannot get session service for killSession. Using closeSession
```

Was I again too slow and timed out ? If not, then this is a bit worrying, never saw it.

**Comment by jburel** [comment]
``%time data = load_from_s3(well.row*well.column-1)``
the time should have been printed

**Comment by pwalczysko** [comment]
I have no luck with Mybinder. It is stuck at "Launching". Trying to refresh, but still, no progress.

Edit: After I have written this, Mybinder moved on.

**Comment by pwalczysko** [comment]
Mybinder including the new commits

```
conn = BlitzGateway(host='ws://idr.openmicroscopy.org/omero-ws', username='public', passwd='public', secure=True)
conn.connect()
conn.c.enableKeepAlive(60)
```

Great that there is the enableKeepAlive now, but this somehow caused the loss of the ``True`` output on execution of that cell.

**Comment by pwalczysko** [comment]
mybinder, with new commits

Evidently after a new build, there are some stale outputs possibly comming from debug testing, such as

```
ClientErrorTraceback (most recent call last)
<ipython-input-20-309636480156> in <module>()
     14     # Load a single Image per Well
     15     image = well.getImage(0)
---> 16     get_ipython().magic(u'time data = load_from_s3(well.index)')
     17     pixels = image.getPrimaryPixels()
     18     size_c = image.getSizeC()

/srv/conda/envs/kernel/lib/python2.
```

I understand these are not true errors, as I did not exeute the cell yet, but still, it is confusing.

Edit: No, this is not a stale output, after execution of the cell, I still have it.

**Comment by pwalczysko** [comment]
Sorry, actually the https://github.com/ome/omero-guide-cellprofiler/pull/9#issuecomment-621132465 have to be modified - I am having this error outputs myself, as I am running the cell 

**Run Cell Profiler Pipeline on the plate**

**Comment by pwalczysko** [comment]
Yes, I can see the timestamps now on 

**Run Cell Profiler Pipeline on the plate**

(I was looking for them in a cell called **Load...** as they were advertised as a speed of loading. We will need to clarify this for demo situations

**Comment by pwalczysko** [comment]
mybinder

After the error https://github.com/ome/omero-guide-cellprofiler/pull/9#issuecomment-621132465 in the cell Run Cell Profiler Pipeline on the plate, all subsequent cells are erroring.

**Comment by jburel** [comment]
I could remove the magic ``%time`` and replace by a print of "end-start".
This is mainly to give us indication during the evaluation phase

**Comment by pwalczysko** [comment]
> I could remove the magic %time and replace by a print of "end-start".

For me it is not what happens in that cell, the problem is evidently not finding the right cell in the first place - the whole overview of the notebook must be helped - but possibly we can do it by non-jupyter means if preferred.

**Comment by pwalczysko** [comment]
This seems like a crash https://github.com/ome/omero-guide-cellprofiler/pull/9#issuecomment-621135594

**Comment by jburel** [comment]
@pwalczysko the problem you noticed is now fixed.

**Comment by pwalczysko** [comment]
repo2docker with the new commits

All as expected, no strange error outputs, no crash, renaming of the cell where you load dask array as expected.

The only thing I am missing now after the keepAlive was added is the confirmation that I am connected - this would be good to have, this did not change since https://github.com/ome/omero-guide-cellprofiler/pull/9#issuecomment-621131381

**Comment by jburel** [comment]
@pwalczysko the last commit will print out ``True`` when connected

**Comment by pwalczysko** [comment]
mybinder

> @pwalczysko the last commit will print out True when connected

Yes, can confirm, the True is displayed now.

The whole experience on mybinder now was good, the notebook makes sense, and (with the exception of mybinder build being very very slow, which is not the fault of this PR), it was also fast (the processing of the 5 wells is definitely faster now).

Good to merge fmpov

**Comment by jburel** [comment]
for info:
using ``s3/dask array``
for one well/image
```
CPU times: user 30 ms, sys: 10 ms, total: 40 ms
Wall time: 285 ms # init time - minio
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 537 µs # channel 1
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 648 µs # channel 2
```
using ``getPlane``
```
CPU times: user 80 ms, sys: 90 ms, total: 170 ms
Wall time: 866 ms # channel 1
CPU times: user 90 ms, sys: 70 ms, total: 160 ms
Wall time: 787 ms # channel 2
```

**Comment by joshmoore** [comment]
To what degree are the ipynb contents and the scripts kept in sync? Does one need to review both of them?

**Comment by jburel** [comment]
They should be in synch. I have pushed an extra commit for the keepAlive
@pwalczysko you may want to run the script in on a couple of wells

**Comment by pwalczysko** [comment]
> @pwalczysko you may want to run the script in on a couple of wells

You mean install this here locally and run the python script called ...zarr ? Not sure I get this.

**Comment by jburel** [comment]
You need to have a conda env as descripted in the setup.py file
The easiest approach is to use ``repo2docker`` section in the ``setup.py``. This can be done in mybinder
When you are in the terminal with the conda env activated, just run ``python scripts/idr0002_zarr.py``

**Comment by pwalczysko** [comment]
Not sure where is the ``setup.py`` file. 

repo2docker, created a new terminal, then went into it (was not sure which conda env to activate, but it does not seem to matter, as it does not work in either ``notebook`` or ``kernel`` envs) 

```
(notebook) pwalczysko@17e1334e7fed:~$ python scripts/idr0002_idr_zarr.py
Traceback (most recent call last):
  File "scripts/idr0002_idr_zarr.py", line 29, in <module>
    from omero.gateway import BlitzGateway
ModuleNotFoundError: No module named 'omero'

(notebook) pwalczysko@17e1334e7fed:~$ conda activate kernel
(kernel) pwalczysko@17e1334e7fed:~$ python scripts/idr0002_idr_zarr.py
Traceback (most recent call last):
  File "scripts/idr0002_idr_zarr.py", line 29, in <module>
    from omero.gateway import BlitzGateway
ModuleNotFoundError: No module named 'omero'
(kernel) pwalczysko@17e1334e7fed:~$
```

**Comment by jburel** [comment]
Sorry ``setup.rst``. All the instructions are in that file about which env to use  etc.
I might have forgotten a commit, checking that now

**Comment by jburel** [comment]
I just run the example and no error.  See below
Screenshot 2020-04-30 at 14 10 41

**Comment by pwalczysko** [comment]
Screen Shot 2020-04-30 at 14 31 50

**Comment by pwalczysko** [comment]
```
(kernel) pwalczysko@04b53c1e9fec:~$ cd scripts/
(kernel) pwalczysko@04b53c1e9fec:~/scripts$ python idr0002_
idr0002_idr_zarr.py        idr0002_save_solution1.py  idr0002_save_solution3.py
idr0002_save.py            idr0002_save_solution2.py
(kernel) pwalczysko@04b53c1e9fec:~/scripts$ python idr0002_idr_zarr.py
/srv/conda/envs/kernel/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Plate ID [422]:
('Remove module: ', 'Images')
('Remove module: ', 'Metadata')
('Remove module: ', 'NamesAndTypes')
('Remove module: ', 'Groups')
Pipeline modules:
(1, 'IdentifyPrimaryObjects')
(2, 'IdentifyPrimaryObjects')
(3, 'RelateObjects')
(4, 'FilterObjects')
(5, 'MeasureObjectIntensity')
(6, 'OverlayOutlines')
(7, 'DisplayDataOnImage')
(8, 'ClassifyObjects')
(9, 'CalculateMath')
(10, 'ExportToSpreadsheet')
analyzing...
plate1_1_013 [Well 88, Field 1 (Spot 88)]
plate1_1_013 [Well 82, Field 1 (Spot 82)]
plate1_1_013 [Well 8, Field 1 (Spot 8)]
plate1_1_013 [Well 19, Field 1 (Spot 19)]
plate1_1_013 [Well 11, Field 1 (Spot 11)]
analysis done
done
(kernel) pwalczysko@04b53c1e9fec:~/scripts$
```
@jburel ok, sorry, now it works - I did not cd into the ``scripts`` directory

**Comment by jburel** [comment]
Thanks @pwalczysko

**Pull request merged**